src/tools/sssctl/sssctl_data.c

```diff
@@ -311,3 +314,222 @@ errno_t sssctl_cache_expire(struct sss_cmdline *cmdline,
+            DEBUG(SSSDBG_CRIT_FAILURE, "Invalid action: %i\n", action);
+            ret = EINVAL;
+            goto done;
+        }
+
+        ret = sysdb_get_db_file(tmp_ctx, NULL, *domain, DB_PATH, &cache, NULL);
+        if (ret != EOK) {
+            DEBUG(SSSDBG_OP_FAILURE, "Failed to get the cache db name\n");
+            goto done;
+        }
+
+        ret = sysdb_manage_index(tmp_ctx, action, cache, attr, &indexes);
+        if (ret != EOK) {
+            goto done;
+        }
+
+        if (action == SYSDB_IDX_LIST) {
+            for (index = indexes; *index != NULL; index++) {
+                PRINT("  Attribute: %1$s\n", *index);
+            }
+            talloc_zfree(indexes);
+        }
+    }
+
+    ret = EOK;
+
+done:
+    talloc_free(tmp_ctx);
+
+    return ret;
+}
+
+errno_t sssctl_cache_index(struct sss_cmdline *cmdline,
+                                  struct sss_tool_ctx *tool_ctx,
+                                  void *pvt)
+{
+    const char *attr = NULL;
+    const char *action_str = NULL;
+    const char **domains = NULL;
+    const char **p;
+    enum sysdb_index_actions action;
+    errno_t ret;
+
+    /* Parse command line. */
+    struct poptOption options[] = {
+        { "domain", 'd', POPT_ARG_ARGV, &domains,
+            0, _("Target a specific domain"), _("domain") },
+        { "attribute", 'a', POPT_ARG_STRING, &attr,
+            0, _("Attribute to index"), _("attribute") },
+        POPT_TABLEEND
+    };
+
+    ret = sss_tool_popt_ex(cmdline, options, SSS_TOOL_OPT_OPTIONAL, NULL, NULL,
+                           "ACTION", "create | delete | list",
+                           SSS_TOOL_OPT_REQUIRED, &action_str, NULL);
+    if (ret != EOK) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "Unable to parse command arguments\n");
+        goto done;
+    }
+
+    if (action_str == NULL) {
+        ERROR("Action not provided\n");
+        ret = EINVAL;
+        goto done;
+    }
+
+    if (strcmp(action_str, "list") == 0) {
+        action = SYSDB_IDX_LIST;
+    } else {
+        if (strcmp(action_str, "create") == 0) {
+            action = SYSDB_IDX_CREATE;
+        } else if (strcmp(action_str, "delete") == 0) {
+            action = SYSDB_IDX_DELETE;
+        } else {
+            ERROR("Unknown action: %1$s\nValid actions are "
+                           "\"%2$s\", \"%3$s and \"%4$s\"\n",
+                  action_str, "create", "delete", "list");
+            ret = EINVAL;
+            goto done;
+        }
+
+        if (attr == NULL) {
+            ERROR("Attribute (-a) not provided\n");
+            ret = EINVAL;
+            goto done;
+        }
+    }
+
+    ret = sssctl_cache_index_action(action, domains, attr);
+    if (ret == ENOENT) {
+        ERROR("Attribute %1$s not indexed.\n", attr);
+        goto done;
+    } if (ret == EEXIST) {
+        ERROR("Attribute %1$s already indexed.\n", attr);
+        goto done;
+    } else if (ret != EOK) {
+        ERROR("Index operation failed.\n");
```

Comment:
Print `ret` to provide more information as to why the action failed.

Reply:
Yes. good idea.